src/fileop.c

```diff
@@ -196,3 +255,195 @@ int librole_get_directory_files(const char * const directory,
 librole_get_directory_files_end:
     return retcode;
 }
+
+/**
+ * \brief Read system role entries from /etc/role.d catalog.
+ *
+ * \param[in] directory directory to open.
+ * \return
+ *  - LIBROLE_INCORRECT_VALUE: Incorrect values passed to function
+ *  - LIBROLE_ERROR_OPENING_DIRECTORY: Error occured while trying to
+ *    open directory.
+ *  - LIBROLE_OK: Operation succeeded.
+ *  - errno: Undefined error.
+ */
+int librole_get_system_roles(const char * const directory,
+    char *system_roles[])
+{
+    int retcode = LIBROLE_OK;
+    struct dirent **files;
+    int file_count = 0;
+    int i, r;
+
+    errno = 0;
+
+    if (NULL == directory || NULL == system_roles)
+    {
+        retcode = LIBROLE_INCORRECT_VALUE;
+        goto librole_get_directory_files_end;
+    }
+
+
+    /* Get all regular files in directory */
+    file_count = scandir(directory, &files, librole_is_role_file, alphasort);
+    if (0 != errno)
+    {
+        retcode = errno;
+        goto librole_get_directory_files_end;
+    }
+
+    for (i = 0, r = 0; i < file_count; ++i)
+    {
+        char* rolename;
+        /* Validate reading filename and skip if name is not valid */
+        if (librole_validate_system_role_filename(files[i]->d_name, &rolename) == LIBROLE_OK) {
+            if (r < LIBROLE_MAX_SYSTEM_ROLES) {
+                system_roles[r++] = rolename;
+            }
+        }
+        free(files[i]);
+    }
+    free(files);
+    system_roles[r] = NULL;
+
+librole_get_directory_files_end:
+    return retcode;
+}
+
+/* delim: 0 - '', > 0 - ',' before, < 0 - ':' after */
+static int write_group(FILE *f, int gid, int delim, int numeric_flag)
+{
+    /* print ',' before if needed */
+    if (delim > 0 && fputc(',', f) < 0)
+        return LIBROLE_IO_ERROR;
+
+    if (numeric_flag) {
+        if (fprintf(f, "%u", gid) < 0)
+            return LIBROLE_IO_ERROR;
+    } else {
+        int result;
```

Comment:
Mixed code and declarations.

Reply:
It is not this pull request subject. This function is already exists.